apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,322 @@
+﻿/*
+ * Copyright 2018 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+using Google.Apis.Auth.OAuth2;
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// Base class for API-specific builders.
+    /// </summary>
+    /// <typeparam name="TBuilder">The type of this builder, for chaining purposes</typeparam>
+    /// <typeparam name="TClient">The type of client created by this builder.</typeparam>
+    public abstract class ClientBuilder<TBuilder, TClient> where TBuilder : ClientBuilder<TBuilder, TClient>
+    {
+        /// <summary>
+        /// The endpoint to connect to, or null to use the default endpoint.
+        /// </summary>
+        public ServiceEndpoint Endpoint { get; private set; }
+
+        /// <summary>
+        /// The scopes to use, or null to use the default scopes.
+        /// </summary>
+        public IReadOnlyList<string> Scopes { get; private set; }
+
+        /// <summary>
+        /// The channel credentials to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public ChannelCredentials ChannelCredentials { get; private set; }
+
+        /// <summary>
+        /// The credential file to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public string CredentialsFile { get; private set; }
+
+        /// <summary>
+        /// The credentials to use as JSON, or null if credentials are being provided in a different way.
+        /// </summary>
+        public string JsonCredentials { get; private set; }
+
+        /// <summary>
+        /// The token access method to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public Func<string, CancellationToken, Task<string>> TokenAccessMethod { get; private set; }
+
+        /// <summary>
+        /// The call invoker to use, or null to create the call invoker when the client is built.
+        /// </summary>
+        public CallInvoker CallInvoker { get; private set; }
+
+        /// <summary>
+        /// Creates a new instance with no settings.
+        /// </summary>
+        protected ClientBuilder()
+        {
+        }
+        
+        /// <summary>
+        /// Creates a clone of the given builder, with respect to settings in this class.
+        /// </summary>
+        /// <param name="builder">The builder to clone</param>
+        protected ClientBuilder(TBuilder builder)
```

Comment:
Is `TokenAccessMethod` missing for a reason?

Reply:
Nope, fixed.